File: Python3/swapping-nodes-in-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        node = head
        count = 0
        while node:
            count += 1
            node = node.next
        k2 = count - k + 1

        node = head
        nodeA = None
        nodeB = None
        i = 1
        while node:
            if i == k:
                nodeA = node
            if i == k2:
                nodeB = node
            i += 1
            node = node.next
            if nodeA and nodeB:
                break
        nodeA.val, nodeB.val = (nodeB.val, nodeA.val)

        return head

File: JavaScript/swapping-nodes-in-a-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var swapNodes = function(head, k) {
    let current = head

    // count elements
    let len = 0
    while (current !== null) {
        current = current.next
        len++
    }

    let first = null
    let second = null

    current = head
    let index = 1
    while (current !== null) {
        if (index == k) {
            first = current
        }
        if (index == len - k + 1) {
            second = current
        }
        current = current.next
        index++
    }

    const tmp = first.val
    first.val = second.val
    second.val = tmp
    return head
};